### ISSUE TITLE
Proposal: add EqualityComparer for ReadOnlyMemory<char> to enable collections with memory backed strings

### ISSUE THREAD

**Issue by cdmihai** [dotnet/corefx#31526]
MSBuild has a bunch of HashSet<string> and Dictionaries with string keys. Due to how Equals and HashCode on ReadOnlyMemory are implemented, it's not possible to convert these collections to have ReadOnlyMemory<char> as keys. An EqualityComparer which derived from the actual view in the string would help here.

```c#
var set = new HashSet<ReadOnlyMemory<char>>(ReadOnlyMemory.StringComparer.OrdinalIgnoreCase);
set.Add("FooBAR".Slice(3));
set.Contains("ZarBar".Slice(3)); // awesome if this returned true
```

**Comment by Wraith2** [comment]
I think this would tie in with https://github.com/dotnet/corefx/issues/31302 in some way but would probably require a specific `ReadOnlyMemory<char>` comparer to be able to take advantage of the extension methods.

There is also a slight worry that while the name ReadOnlyMemory claims to be readonly it isn't actually guaranteeing anything about the underlying storage being immutable. String instances are (within managed bounds) immutable so using them as keys is ok, can you be certain that any `ReadOnlyMemory<T>` wrapping some object won't be mutated elsewhere?

**Comment by benaadams** [comment]
`ReadOnlyMemory` as actual keys would perform badly and be problematic (wrapping mutable memory, holding reference to entire object just for a slice, lifetime issues)

However maybe as an extension for lookups on a `string` dictionary or hashset? Though Span would likely be more flexible at that point...
```csharp
public static CollectionExtensions<TValue>
{
    public bool TryGetValue(this Dictionary<string, TValue> dict, ReadOnlySpan<char> key, out TValue value);
    public bool ContainsKey(this Dictionary<string, TValue> dict, ReadOnlySpan<char> key);
}

public static CollectionExtensions
{
    public bool TryGetValue(this HashSet<string> dict, ReadOnlySpan<char> equalValue, out string actualValue);
    public bool Contains(this HashSet<string> dict, ReadOnlySpan<char> key);
}
```

**Comment by Wraith2** [comment]
So those methods would take the span convert to a string and then pass it to the dictionary? Unless the string borrowed the underlying storage that'd require allocation of a new string each time which doesn't seem desirable.

**Comment by benaadams** [comment]
> So those methods would take the span convert to a string and then pass it to the dictionary? 

No they'd need to be tied to the implementation; but they'd also only be valid for `char` types. Would need to use the same hashcode as the string used in the Collection and also same comparer

**Comment by Wraith2** [comment]
I don't think that would work. The keys are stored in the entries in the dictionary so there do actually have to be real strings unless you change TKey to be something more generic like IEnumerable<char>.

**Comment by benaadams** [comment]
The keys in the Dictionary would be strings; doesn't mean the values used to look up have to be strings; they could be ReadOnlySpan.

e.g. use case an intern Hashtable that returns strings for a ROS is the string has already been created

**Comment by Wraith2** [comment]
In [Dictionary.FindEntry](https://github.com/dotnet/coreclr/blob/5164e76ce01fc36833a7f5e6fb7b1ec2f55c2262/src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs#L415) the stored and lookup keys are passed to the provided comparer. Both must be of type TKey for IEqualityComparer.Equals and string is sealed so we can't derive our way out. Is there a way around this that I can't see?

**Comment by benaadams** [comment]
> In the end, TKey is reduced to an int via GetHashCode, so maybe Dictionary and HashSet could have an int based lookup?

It also needs equality testing; the hashcode `int` is a candidate check (in addition to bucket location) can be more than one entry with the same hashcode

**Comment by cdmihai** [comment]
Good point, forgot about collisions. Stack confined lambdas would help here, but that would need language modifications:

```c#
public HashSet<T>
{
    bool TryGetValue(int hashcode, ref Func<T, bool> equalityChecker, out T actualValue);
}
```

```c#
var set = new HashSet<string>(){"foo", "bar"};
var slice = "fooBAR".Slice(3);
set.TryGetValue(String.GetHashCode(slice, StringComparison.OrdinalIgnoreCase), ref s => slice.Equals(s.AsSpan(), StringComparison.OrdinalIgnoreCase), out var value);
```
An alternative to ref lambdas would be introducing a structural typing idiom like foreach and Enumerable which would lookup Equals and GetHashCode on structs. This way, one would pass an IEqualityComparer looking struct to the lookup collection.

**Comment by Wraith2** [comment]
Using existing features we could have a new type `StringLikeKey` which contains a reference to something that's string like and can compare that to other string like things. The performance on comparison won't be perfect because there will be type detection but actual comparison is hopefully a rare event. The GetHashCode result can be cached in it which would slightly mitigate underlying mutation. it'll should probably be a class because we know we're going to store it in a bucket and there's no point copying it about. Sprinkle in some implicit conversions and it might work.

The original comment by @benaadams on perf would still stand though, if you add something to `Dictionary<StringLikeKey,TValue>` with a `ReadOnlyMemory<char>` as the backing value you've got to be careful to clear out that dictionary or risk causing problems. I can see it being a useful tool when parsing though, pull in a file as a single blob use it in lookups without reallocating pieces of it as strings and then discard the entire set of related structures.

The slicing string hashcode overload you mention would be dealt with my the hashcode extension for `ReadOnlySpan<char>` in https://github.com/dotnet/corefx/issues/31302 , you just slice to span and hash that.

When doing xml parsing I've hit this same sort of issue. `NameTable` can be used to lookup common tokens but I can't use it with span without going to an `int` key and losing the collision capability.

As useful as span I always seem to end up doing something slightly annoying at the intersection of span to string based code. It's a bit like going from generic to non-generic collections, slightly jarring.

**Comment by benaadams** [comment]
Might be better as a specific feature of a "Power Collection" https://github.com/dotnet/corefxlab/issues/2406 e.g. where the key is a string-type rather than trying to get the generic collections to work with this caveat for string-types?

**Comment by KrzysztofCwalina** [comment]
Can we just implement `IEqualityComparer<Memory<char>>` on StringComparer?
